extensions/common/gcp/gcp-core/src/main/java/org/eclipse/edc/gcp/bigquery/BigQueryServiceImpl.java

```diff
@@ -0,0 +1,594 @@
+                } finally {
+                    try {
+                        outputStream.close();
+                        monitor.debug("BigQuery Source output stream closed");
+                    } catch (IOException ioException) {
+                        monitor.severe("BigQuery Source exception closing the output stream", ioException);
+                    }
+                }
+            }).start();
+
+            return parts.stream();
+        } catch (IOException ioException) {
+            throw new GcpException(ioException);
+        }
+    }
+
+    private void append(CircularBuffer buffer)
+        throws DescriptorValidationException, IOException, InterruptedException {
+        while (buffer.getAvailableReadCount() > 0) {
+            var content = buffer.getString();
+            if (!content.startsWith("[")) {
+                throw new IOException("BigQuery Sink invalid JSON Array '" + content + "'");
+            }
+
+            int endIndex = content.indexOf(']');
+            if (endIndex != -1) {
+                var pageString = content.substring(0, endIndex + 1);
+                var page = new JSONArray(pageString);
+                // TODO re-try the append operation for a maximum, specified number of times.
+                checkStreamWriter();
+                ApiFuture<AppendRowsResponse> future = streamWriter.append(page);
+                ApiFutures.addCallback(
+                        future, new AppendCompleteCallback(), MoreExecutors.directExecutor());
+
+                buffer.read(pageString.length());
+            } else {
+                return;
+            }
+        }
+    }
+
+    private boolean closeStream() {
+        try {
+            checkStreamWriter();
+            streamWriter.close();
+            return true;
+        } catch (IOException | InterruptedException | DescriptorValidationException exception) {
+            monitor.severe("BigQuery Sink error while closing the stream writer", exception);
+        }
+
+        return false;
+    }
+
+    @Override
+    public void runSinkQuery(List<DataSource.Part> parts) {
+        if (parts.isEmpty()) {
+            return;
+        }
+
+        CircularBuffer buffer = new CircularBuffer(16384);
+        var errorWhileAppending = false;
+        for (var part : parts) {
+            buffer.reset();
+            try (var inputStream = part.openStream()) {
+                boolean done = false;
+                while (!done) {
+                    var available = buffer.getAvailableWriteCount();
+                    var dataRead = inputStream.read(buffer.getBuffer(), buffer.getWriteOffset(), available);
+                    if (dataRead > 0) {
+                        buffer.write(dataRead);
+                        append(buffer);
+                    } else if (dataRead == -1) {
+                        done = true;
+                    } else if (dataRead == 0) {
```

Comment:
at this point dataRead will always be 0 (by definition it can't be less that `-1`

Reply:
Done